### DISCUSSION TITLE
How to download twitch streams that use av1/h265 codec

### DISCUSSION THREAD

**Discussion by wolfgangam31** [streamlink/streamlink#6356] [Q&A] [answered]
### Description

HI everyone, I wanted to ask about the new twitch codec that some streamers started to use a few months ago, 
Here for example - https://www.twitch.tv/videos/2335165116 

when using streamlink to download these, I  get a file but it's barely usable... most players don't play it all, windows "movies & tv" plays it but I can't use the timeline, and MPV plays it as well, but it takes about a minute to load , but once it loads it does seem to at least work fine I think. 

When I look at the chunks being downloaded in chrome network tab, they are now mp4, and if I download one to my pc, I can't watch it in VLC / mpv / windows media player, I also can't seem to get any info on it with ffprobe... why is that? 

How can I work with these files ? or can I somehow get the older TS links for these streams?

**Comment by bastimeyer** [comment, marked as answer]
https://github.com/streamlink/streamlink/pull/5769
https://github.com/streamlink/streamlink/pull/5769#issuecomment-2543139167

**Reply by wolfgangam31** [reply]
Hi, thanks, I read all of it prior to opening this, I'm sorry to say I didn't fully understood what is going on there, so I opened this . 

When trying to search anything about the command there "twitch-supported-codecs" gave me no results , and when trying to use it, it seems it's not valid command anymore?

I saw the comment about PotPlayer , but their main website doesn't currently have HTTS , so I opted out of downloading it at the moment. 
I would still love to understand, just as a discussion, why simply downloading the mp4 from the network tab or from m3u8 file , I can't read them in ffprobe for example or play them ...

**Reply by bastimeyer** [reply]
This is an **unmerged** in-development pull request.
If you want to use the newly added functionality, then you'll have to build Streamlink from the branch of that pull request.

- https://streamlink.github.io/install.html#pypi-package-and-source-code
- https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#pull-request-feedback

The recent commits on that PR branch automatically remux the MPEG-4 container that's used for AV1 and h265 video codecs into an MPEG-TS transport stream, just like the regular h264 streams on Twitch. The presentation timestamps are also set to 0 at the point where Streamlink started downloading in order to prevent player confusion, since the timestamps in the used MPEG-4 container reference the start of the broadcast.

If you're using direct HLS playlist URLs as player input, then you're not using Streamlink. Streamlink has its own HLS client implementation.

You can't read single "mp4 files" from an HLS stream, because they require an initialization segment, which is included in the HLS playlist metadata. `cat init.mp4 segment.mp4 | ffprobe -i pipe:0 ...` works.

**Reply by wolfgangam31** [reply]
again, thanks for explanation. So with the current stable version, should I be able to download both vods and also live streams that use the new encoder ? I could swear that when I tried to download live one, the file at the end was not usable, or it might have been audio only?, but later downloaded the vod , the file did work, does it make since or I just need to try again next time I get a chance to catch a live stream using it?

about the HLS, yes I was checking the playlist without streamlink to try and understand what is going on, just out of curiosity... 
I tried the command you gave for ffprobe, so yes I do see the details now, and I also downloaded now init and "cat" it with a segment and the file I downloaded did open, for some reason it was just audio (like what I think happened when downloading live with streamlink), but I don't remember where If I got that random segment from live stream or vod (it was just sitting in my download folder) ...  I redownloaded a fresh one, just to check , and it seems fine,  with a video.

**Reply by bastimeyer** [reply]
> So with the current stable version, should I be able to download both vods and also live streams that use the new encoder ?

No, as explained, this is an unmerged in-development branch. I've linked all the information you need in order to install from the PR branch. Good luck.

Streamlink 7.0.0 or builds from the current HEAD of the `master` branch do not support Twitch's AV1/h265 streams, as they need to be requested specifically on Twitch's usher API (where HLS playlist URLs are retrieved from). Without requesting the newly supported video codecs, Twitch will only return a set of re-encoded h264 streams in a regular transport stream.

This is also the reason why this hasn't been merged yet since January when the PR was opened, because this might change at some point. It's still considered beta status by Twitch. They could also include all available streams in the multivariant playlist, requiring the HLS client to filter the streams based on playback capability.

Once again, all the information is already out there, so there's no need to repeat this again.
